src/desktop/usb.rs

```diff
@@ -0,0 +1,282 @@
+// Copyright (C) 2024 GNOME Foundation
+//
+// Authors:
+//     Hubert Figuière <hub@figuiere.net>
+//
+
+//! Provide an interface to USB device. Allow enumerating devices
+//! and requiring access to.
+//! ```rust,no_run
+//! use ashpd::desktop::usb::UsbProxy;
+//! use futures_util::StreamExt;
+//!
+//! async fn watch_devices() -> ashpd::Result<()> {
+//!     let usb = UsbProxy::new().await?;
+//!     let session = usb.create_session().await?;
+//!     if let Some(response) = usb.receive_device_events().await?.next().await {
+//!         let events = response.events();
+//!         for ev in events {
+//!             println!(
+//!                 "Received event: {} for device {}",
+//!                 ev.event_action(),
+//!                 ev.event_device_id()
+//!             );
+//!         }
+//!     }
+//!
+//!     Ok(())
+//! }
+//! ```
+
+use std::collections::HashMap;
+
+use futures_util::Stream;
+use serde::Deserialize;
+use zbus::zvariant::{
+    DeserializeDict, ObjectPath, OwnedFd, OwnedObjectPath, OwnedValue, SerializeDict, Type, Value,
+};
+
+use crate::{
+    desktop::{HandleToken, Session, SessionPortal},
+    proxy::Proxy,
+    Error, WindowIdentifier,
+};
+
+#[derive(Debug, SerializeDict, Type, Default)]
+#[zvariant(signature = "dict")]
+struct CreateSessionOptions {
+    handle_token: HandleToken,
+    session_handle_token: HandleToken,
+}
+
+/// Options for the USB portal.
+#[derive(SerializeDict, Type, Debug, Default)]
+#[zvariant(signature = "dict")]
+struct UsbEnumerateOptions {}
+
+/// USB device description
+#[derive(SerializeDict, DeserializeDict, Type, Debug, Default)]
+#[zvariant(signature = "dict")]
+pub struct UsbDevice {
+    parent: Option<String>,
+    /// Device can be opened for reading. Default is false.
+    readable: Option<bool>,
+    /// Device can be opened for writing. Default is false.
+    pub writable: Option<bool>,
+    /// The device node for the USB.
+    #[zvariant(rename = "device-file")]
+    pub device_file: Option<String>,
+    /// Device properties
+    pub properties: Option<HashMap<String, OwnedValue>>,
+}
```

Comment:
The way we deal with structs in general is that no field is public, only add getters or setters that would be public. This allow us to change the way we deserialize/serialize things without impacting the public API.